src/OrchardCore.Modules/OrchardCore.Sitemaps/Builders/CustomPathSitemapSourceBuilder.cs

```diff
@@ -0,0 +1,78 @@
+using System.Globalization;
+using System.Threading.Tasks;
+using System.Xml.Linq;
+using Microsoft.AspNetCore.Mvc;
+using OrchardCore.Sitemaps.Models;
+using OrchardCore.Mvc.Core.Utilities;
+
+namespace OrchardCore.Sitemaps.Builders
+{
+    public class CustomPathSitemapSourceBuilder : SitemapSourceBuilderBase<CustomPathSitemapSource>
+    {
+        private static readonly XNamespace Namespace = "http://www.sitemaps.org/schemas/sitemap/0.9";
+
+        public override async Task BuildSourceAsync(CustomPathSitemapSource source, SitemapBuilderContext context)
+        {
+            var url = new XElement(Namespace + "url");
+
+            if (await BuildUrlsetMetadataAsync(source, context, url))
+            {
+                context.Response.ResponseElement.Add(url);
+            }
+        }
+
+        private Task<bool> BuildUrlsetMetadataAsync(CustomPathSitemapSource source, SitemapBuilderContext context, XElement url)
+        {
+            if (BuildUrl(context, source, url))
+            {
+                PopulateLastMod(source, url);
+                PopulateChangeFrequencyPriority(source, url);
+
+                return Task.FromResult(true);
+            }
+
+            return Task.FromResult(false);
+        }
+
+        private bool BuildUrl(SitemapBuilderContext context, CustomPathSitemapSource source, XElement url)
+        {
+            if (string.IsNullOrEmpty(source.Url))
+                return false;
+
+            if(!(context.HostPrefix?? "").EndsWith("/") && !source.Url.StartsWith("/") && !source.Url.StartsWith("~"))
```

Comment:
If you assume `blog/post` no need for extra checks here. Just prepend `~/` virtual path to all I suspect